### PR TITLE
plugins/todo-comments: migrate to mkNeovimPlugin

### PR DESCRIPTION
**Pull request by khaneliman**

### CODE REVIEW

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:268]: I think transition type would be confused by the composite type (listOf). Especially for empty list definitions.

Perhaps we could do `either (transitionType str) (listOf (transitionType str))`?

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:270]: Just to clarify, `[[ ]]` is just a string syntax in lua (raw strings to be precise), anything that is written using `[[ ]]` can also be written using `" "` or `' '`, so I think this option can actually just be `either str (listOf str)` and the old option can be removed if it's incompatible?

Or if we want to be graceful the old option could be kept and we could copy it's definition to here when it's defined.

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:342]: Same. This doesn't need to be a raw strings.

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:255]: This should be effectively equivalent to `toLuaObject cfg.search.pattern`

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:362]: Is it possible to do this is a way that complies with https://github.com/nix-community/nixvim/issues/603?

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:408]: I see this is unchanged from the previous implementation, other than being indented further.

However, I think this would be simpler mapping over `cfg.keymaps` instead of mapping over `commands`?

Ideally, I'd like to avoid using `flatten`, perhaps we could do something like:

```nix
keymaps = lib.pipe cfg.keymaps [
  (filterAttrs (n: v: v != null))
  (mapAttrsToList (name: keymap: {
    inherit (keymap) key option;
    mode = "n";
    action =
      let
        cmd = commands.${name};
        cwd = optionalString (keymap.cwd != null) " cwd=${keymap.cwd}";
        keywords = optionalString (keymap.keywords != null) " keywords=${keymap.keywords}";
      in
      "<cmd>${cmd}${cwd}${keywords}<cr>";
  }))
];
```

Looking at the `commands` attrset, we could probably replace it with a `upperFirstChar` style helper.

However, many of the details would be slightly different if we can go down the #603 route.

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:414]: Is this something we still want to do, or would we rather warn/assert asking the user to enable the plugins themselves?

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [reply plugins/utils/todo-comments.nix:356]: If you're unfamiliar, the idea is that plugin keymaps should make use of `mkMapOptionSubmodule` instead of implementing their own keymap submodule types.

Comment (MattSturgeon) [reply plugins/utils/todo-comments.nix:356]: If `mkMapOptionSubmodule` is too limiting, because it doesn't support additional options (`cwd`, `keywords`) I'm open to making it more powerful by adding an `extraOptions` argument or similar.

File: plugins/utils/todo-comments.nix
Comment (khaneliman) [reply plugins/utils/todo-comments.nix:414]: This is something I saw and mentioned in https://github.com/nix-community/nixvim/pull/1559. I wasn't sure what everyone's opinions were so I left it untouched. But, based on that PR it would seem we'd want to change it to assertions.

File: plugins/utils/todo-comments.nix
Comment (khaneliman) [reply plugins/utils/todo-comments.nix:408]: > I see this is unchanged from the previous implementation, other than being indented further.
> 
Yeah, sorry I didn't touch this bit. I wasn't sure what to do with it, yet.

I think the suggested solution is a lot cleaner.  But, I need to get more familiar with that keymap issue that was mentioned here and above.

Comment (MattSturgeon) [reply plugins/utils/todo-comments.nix:408]: We can always tackle it in another PR, but the other suggestions above could be better.

Comment (khaneliman) [reply plugins/utils/todo-comments.nix]: If I can tackle that separately it would give me a bit to figure it out and maybe get the extraOptions support for mkMapOptionSubmodule

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [reply plugins/utils/todo-comments.nix:353]: As discussed in https://github.com/nix-community/nixvim/pull/2030#discussion_r1720865792 this can done in be another PR

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:22]: Up to you, we could drop this, since we have `lib.nixvim.upperFirstChar`: https://github.com/nix-community/nixvim/blob/78fc4be6a830e8dc01f3e66ddbe3243b4bfe8560/lib/utils.nix#L81-L104

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:420]: _If_ you decide to drop the `keymapsActions` attrs:
```suggestion
              cmd = lib.nixvim.upperFirstChar name;
```

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:392]: Instead of null-checking each step, we can use the fallback operator:

```suggestion
          cfg.keymaps.todoTelescope.key or null != null
```

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:401]: Same:
```suggestion
          cfg.keymaps.todoTrouble.key or null != null
```

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:395]:
```suggestion
          Nixvim(plugins.todo-comments): You have enabled todo-comment's `telescope` integration.
```

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:404]:
```suggestion
          Nixvim(plugins.todo-comments): You have enabled todo-comment's `trouble` integration.
```

Comment (MattSturgeon) [reply plugins/utils/todo-comments.nix:404]: IDK if the `(plugins.todo-comments)` is really needed since it's obvious from the assertion text, be we _usually_ include something like it in our warnings/assertions?

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:375]: This seems kinda backwards, since this is a manually crafted option why don't we just use a `listOf str` type and then `concatStringsSep ","` ourselves later? It's not like this is a representation of a freeform settings option.

We could use `transitionType` here, however maybe it's best to wait until we migrate to `mkMapOptionSubmodule`, in case we need other deprecations.

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [reply plugins/utils/todo-comments.nix:22]: Well, I guess it's used in the option declaration as well as the keymap transformation. So maybe best to keep it?

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:43]: This is the same as `mkNeovimPlugin`'s `deprecateExtraOptions` option:
https://github.com/nix-community/nixvim/blob/78fc4be6a830e8dc01f3e66ddbe3243b4bfe8560/lib/neovim-plugin.nix#L25

I don't mind if you'd rather do it explicitly, but it might be easier to keep the dated `TODO` comments separate if we move this out of imports.

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:42]: This was added 2023-11-06 in e45be89f1267faaae0e67fd6d3f3a2457314a94c

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [reply plugins/utils/todo-comments.nix:43]: The imports can be dated 2023-11-06 while `deprecateExtraOptions` and `optionsRenamedToSettings` can be grouped together and dated `2024-08-18`.

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:146]: Note: this will be affected by #1618, however that doesn't affect how you should implement it here. It's just another reason for us to investigate the issue.

Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:146]: Should this have a freeform type? Or should we restrict users to only these sub-options?

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:248]: Looks like maybe an enum?

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:254]: Same (enum?)

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:258]: Same (enum?)

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
Think these are the final comments :rocket:

File: plugins/utils/todo-comments.nix
Comment (khaneliman) [reply plugins/utils/todo-comments.nix:392]: Thanks, I was hoping this could be simplified with some kind of null conditional operator.

File: plugins/utils/todo-comments.nix
Comment (khaneliman) [reply plugins/utils/todo-comments.nix:395]: lol fun copy pasta

File: plugins/utils/todo-comments.nix
Comment (khaneliman) [reply plugins/utils/todo-comments.nix:142]: I'm torn on it, these are the kinds of things that are limited to options and it can be nice to have available in documentation. I think it will be easier to remove the maintenance when we figure out the documentation only options.

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:391]: I don't _think_ we need the null check in the coercion function?

If the value is null then it'll fail `types.str.check` and will end up being merged using the `nullOr` type anyway.

```suggestion
                transitionType str (splitString ",") (
```

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:444]:
```suggestion
                keymap.keywords != null && keymap.keywords != [ ]
```

Or is it intentional to be able to pass an empty list?

If not, maybe we can get rid of the `nullOr` (moving it to the left of the `transitionType`)?

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [reply plugins/utils/todo-comments.nix:142]: I meant having `freeformType = attrsOf anything` in addition to `options = {}`.

We don't need to do this, if you want to restrict the user to only the sub-options declared, but if you don't want that restriction, we need to define a freeformType.

File: plugins/utils/todo-comments.nix
Comment (MattSturgeon) [inline plugins/utils/todo-comments.nix:423]:
```suggestion
        assertion = cfg.keymaps.todoTrouble.key or null != null -> config.plugins.trouble.enable;
```

Typo

File: plugins/utils/todo-comments.nix
Comment (khaneliman) [reply plugins/utils/todo-comments.nix]: It was required when i was doing transition with null old type, then I realized setting null as an old type caused it to trigger even when it wasn't needed. Now it's only attempting to coerce properly and it can be removed.

File: plugins/utils/todo-comments.nix
Comment (khaneliman) [reply plugins/utils/todo-comments.nix]: Not sure why someone would choose to pass an empty list, it should be the same effect as null to the plugin so it makes sense to add it to the condition to avoid trying to concat an empty list.

File: plugins/utils/todo-comments.nix
Comment (khaneliman) [reply plugins/utils/todo-comments.nix]: That makes sense, I was thinking it replaced the existing options for some reason.

**Review by MattSturgeon — APPROVED** [review]
Can't see any remaining issues. Thanks for another quality PR! :tada: 

Using `mkMapOptionSubmodule` as per #603 can be postponed to another PR, and may be best done after #2035 is merged anyway.

Up to you if you're happy merging now if if you'd like to ping other maintainers for a second opinion.

No need to wait for the tests to pass though, since mergify will do that for us.

**Comment by khaneliman** [comment]
@mergifyio queue

**Comment by MattSturgeon** [comment]
@mergifyio refresh

**Comment by MattSturgeon** [comment]
@mergifyio queue

**Pull request merged**
